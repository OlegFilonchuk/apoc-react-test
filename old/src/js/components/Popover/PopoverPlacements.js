import invert from 'lodash/invert';

const POPOVER_PLACEMENTS = {
  LEFT_TOP: 'leftTop',
  RIGHT_TOP: 'rightTop',
  TOP_RIGHT: 'topRight',
  TOP_LEFT: 'topLeft',
  BOTTOM_RIGHT: 'bottomRight',
  BOTTOM_LEFT: 'bottomLeft',
  TOP_MIDDLE: 'topMiddle',
  RIGHT_MIDDLE: 'rightMiddle',
  LEFT_MIDDLE: 'leftMiddle',
  BOTTOM_MIDDLE: 'bottomMiddle'
};

const oppositePlacementsPartial = {
  [POPOVER_PLACEMENTS.BOTTOM_RIGHT]: POPOVER_PLACEMENTS.BOTTOM_LEFT,
  [POPOVER_PLACEMENTS.TOP_RIGHT]: POPOVER_PLACEMENTS.TOP_LEFT,
  [POPOVER_PLACEMENTS.TOP_MIDDLE]: POPOVER_PLACEMENTS.BOTTOM_MIDDLE,
  [POPOVER_PLACEMENTS.RIGHT_MIDDLE]: POPOVER_PLACEMENTS.LEFT_MIDDLE
};

export const POPOVER_PLACEMENTS_OPPOSITE = {
  ...oppositePlacementsPartial,
  ...invert(oppositePlacementsPartial)
};

Object.freeze(POPOVER_PLACEMENTS);

export default POPOVER_PLACEMENTS;
