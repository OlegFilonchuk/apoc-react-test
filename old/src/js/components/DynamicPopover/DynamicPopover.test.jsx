import React from 'react';
import { mount } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';

import PopoverPlacements, { POPOVER_PLACEMENTS_OPPOSITE } from './../Popover/PopoverPlacements';

describe('DynamicPopover', () => {
  let errorConsole;
  let component;

  beforeAll(() => {
    errorConsole = spyOn(console, 'error');
  });

  describe('with default props', () => {
    /* eslint-disable global-require */
    jest.resetModules();
    const DynamicPopover = require('./DynamicPopover').default;
    /* eslint-enable global-require */

    beforeEach(() => {
      component = mount(<DynamicPopover>Some content</DynamicPopover>);
    });

    it('matches snapshot', () => {
      expect(renderToJson(component)).toMatchSnapshot();
    });
  });

  describe('stateUpdaterToAvoidViewportOverflow', () => {
    /**
     * @type {onVisibilityChangeParams}
     */
    const data = {
      isVisible: true,
      tooltipX: 0,
      tooltipWidth: 0,
      contentWidth: 0,
      contentParentWidth: 0
    };

    const initialPlacement = PopoverPlacements.LEFT_TOP;
    const oppositePlacement = POPOVER_PLACEMENTS_OPPOSITE[initialPlacement];
    let DynamicPopover;

    beforeEach(() => {
      component = mount(<DynamicPopover initialPlacement={initialPlacement}>Some content</DynamicPopover>);
    });

    describe('overflowsEdge returns true and calcTooltipWidth returns 100', () => {
      beforeAll(() => {
        /* eslint-disable global-require */
        jest.resetModules();
        jest.doMock('./overflowsEdge', () => () => true);
        jest.doMock('./calcTooltipWidth', () => () => 100);
        DynamicPopover = require('./DynamicPopover').default;
        /* eslint-enable global-require */
      });

      it('its state updater returns proper state change set', () => {
        const stateUpdater = component.instance().stateUpdaterToAvoidViewportOverflow(data);
        const stateChangeSet = stateUpdater(component.state);

        expect(stateChangeSet).toEqual({
          popoverStyles: {
            width: 100,
            maxWidth: 100
          },
          placement: oppositePlacement
        });
      });
    });

    describe('overflowsEdge returns false and calcTooltipWidth returns 100', () => {
      beforeAll(() => {
        /* eslint-disable global-require */
        jest.resetModules();
        jest.doMock('./overflowsEdge', () => () => false);
        jest.doMock('./calcTooltipWidth', () => () => 100);
        DynamicPopover = require('./DynamicPopover').default;
        /* eslint-enable global-require */
      });

      it('its state updater returns proper state change set', () => {
        const stateUpdater = component.instance().stateUpdaterToAvoidViewportOverflow(data);
        const stateChangeSet = stateUpdater(component.state);

        expect(stateChangeSet).toEqual({
          popoverStyles: {
            width: 100,
            maxWidth: 100
          },
          placement: initialPlacement
        });
      });
    });
  });

  describe('onVisibilityChange', () => {
    beforeAll(() => {
      /* eslint-disable global-require */
      jest.resetModules();
      const DynamicPopover = require('./DynamicPopover').default;
      /* eslint-enable global-require */

      component = mount(<DynamicPopover>Some content</DynamicPopover>);
      component.instance().stateUpdaterToAvoidViewportOverflow = jest.fn();
    });

    beforeEach(() => {
      component.instance().stateUpdaterToAvoidViewportOverflow.mockReset();
    });

    describe('data.isVisible = false', () => {
      it('stateUpdaterToAvoidViewportOverflow is not executed', () => {
        const data = { isVisible: false };

        component.instance().onVisibilityChange(data);
        expect(component.instance().stateUpdaterToAvoidViewportOverflow).not.toBeCalled();
      });
    });

    describe('data.isVisible = true', () => {
      it('stateUpdaterToAvoidViewportOverflow is executed', () => {
        const data = { isVisible: true };

        component.instance().onVisibilityChange(data);
        expect(component.instance().stateUpdaterToAvoidViewportOverflow).toBeCalledWith(data);
      });
    });
  });

  it('prints no errors from propTypes validation', () => {
    expect(errorConsole).not.toHaveBeenCalled();
  });
});
