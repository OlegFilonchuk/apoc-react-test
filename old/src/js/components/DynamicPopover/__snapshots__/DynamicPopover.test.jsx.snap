// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicPopover with default props matches snapshot 1`] = `null`;
